### ISSUE TITLE
May I ask why the source vocab and target vocab are separated?

### ISSUE THREAD

**Issue by loveritsu929** [moonlightlane/QG-Net#13]
It seems a bit late but I just read your paper and tried to run the demo. Amazing work, thank you for sharing the repo.

![image](https://user-images.githubusercontent.com/20154344/87371797-0e88ea00-c554-11ea-8a87-b7ea71926c56.png)
I take a look at the model structure and find that encoder and decoder have their own embedding layers of different sizes, meaning the source and target vocabularies are separated. It makes sense if it's the default setting in OpenNMT since the source and target are in two languages in translation tasks. However, here we are dealing with the QG task where both of them are English sentences, similar to the summarization task. 

In Abi's original implementation we can see that the embeddings are shared by encoder and decoder (https://github.com/abisee/pointer-generator/blob/master/model.py @ line 209).

I'm wondering how the copy mechanism works exactly here. If it's the case in the paper that the generation vocab for ith question is the union of the question vocabulary and the vocabulary of the ith input context, does that mean the purpose of using pointer network is not to deal with OOVs? because in the "extended generation vocab" for each question, you are adding words that are in the encoder's vocab but are not in the decoder's vocab, then they are not really OOVs for the entire model. 

Second, in this setting, whenever there is a copy, that means the word is not in the question vocabulary, so there will be a mismatch between the reference question and generated question. To me, it seems that the loss will increase every time the copy mechanism works.

I'm not sure if I have a misunderstanding about your work. Please correct me if you find any. I will appreciate your reply.

**Comment by zichaow** [comment]
Hi, thanks for your interest in the work!

- why source and target has different vocab: our assumption is that questions usually contains a much more limited set of vocabulary than the input source context. This constrains the vocabulary that the generator can use to generate a question, which I think makes sense in the context of question generation.

- the role of pointer network (or copy mechanism): this is to twofold. First, it deals with OOV words. This is natural because the source and target vocab are different; in our context the target vocab will be much smaller than the source vocab so I think it is natural to copy words from the source. Second, copying words from the source also makes the question more relevant to the input, i.e., it is a way to make sure the question is not irrelevant to the input. 

- copy mechanism and OOV: When copy happens, it does not necessarily mean a word not in the question vocab appears in the generated question. For a word that appears in both the question sentence and input sentence(s), copy mechanism simply increases the probability of this word. Of course, during training, there will be no OOV words in the generation because the model has full knowledge of what word will appear in the question. For testing, however, copy mechanism does have the capability to generate a word that appears in the input but not in the question sentence. 

I hope the above answers some of your questions. Please let me know if I can provide further clarifications. Thanks!

**Comment by loveritsu929** [comment]
> Hi, thanks for your interest in the work!
> 
> * why source and target has different vocab: our assumption is that questions usually contains a much more limited set of vocabulary than the input source context. This constrains the vocabulary that the generator can use to generate a question, which I think makes sense in the context of question generation.
> * the role of pointer network (or copy mechanism): this is to twofold. First, it deals with OOV words. This is natural because the source and target vocab are different; in our context the target vocab will be much smaller than the source vocab so I think it is natural to copy words from the source. Second, copying words from the source also makes the question more relevant to the input, i.e., it is a way to make sure the question is not irrelevant to the input.
> * copy mechanism and OOV: When copy happens, it does not necessarily mean a word not in the question vocab appears in the generated question. For a word that appears in both the question sentence and input sentence(s), copy mechanism simply increases the probability of this word. Of course, during training, there will be no OOV words in the generation because the model has full knowledge of what word will appear in the question. For testing, however, copy mechanism does have the capability to generate a word that appears in the input but not in the question sentence.
> 
> I hope the above answers some of your questions. Please let me know if I can provide further clarifications. Thanks!

Hi, thanks for the quick response! 

I just dig into the implementation a bit. Could you provide more explanation of "For a word that appears in both the question sentence and input sentence(s), copy mechanism simply increases the probability of this word."? It seems that all words in the ith source are added to the generation vocabulary ( _dynamic_dict() in TextDataset & src_map) for ith question.

I notice the CopyGenerator returns torch.cat([output_prob, copy_prob]). I don't check the sizes of these two probabilities, but I assume that output_prob is of size (question_vocab) and copy_prob is of size (ith_input_vocab). So, will the words that are in both input and question appear twice in this concatenation?

**Comment by zichaow** [comment]
Hi ! I think what happens (on a high level) is that a new vocab is created in the process. Here is a simple example:

input: new york is a city in usa .
question: where is new york located ?

The original vocab for this question: ["where", "is", "new", "york", "located"]

let's say we are generating the third word. Ideally, we want to generate the word "new".

Without copy mechanism, the following could be the softmax output (bad example but illustrative):
["where": 0.2, 
"is":0.2, 
"new":0.2,
"york":0.2, 
"located":0.2]

With copy mechanism, we do three things here:
1) get the attention weights for the words in the input:
["new": 0.3,  
"york": 0.1, 
"is": 0.1, 
"a": 0.1, 
"city": 0.1, 
"in": 0.1, 
"usa": 0.2]
2) compute the copy probability, say it is 0.8, which means we prefer to copy a word from the input to the generated question
3) add the two lists together, weighted by the copy probability. So for example, the word "where" gets a probability of 0.2 * (1-0.8) + 0 * 0.8 = 0.04 (weighted by the not-copy probability which is 1-0.8=0.2, and add 0 * 0.2 because it does not appear in the input). Similarly, the word "new" gets a probability of 0.2 * (1-0.8) + 0.3 * 0.8 = 0.28. The other word probabilities can be computed similarly. 

Carrying the above process for all words in the input and question, you'll see that the word "new" has the highest probability among the augmented vocabulary (vocab of question + vocab of input) and thus we select this word. 

It has been some time since I update the repo but I can take a look at the implementation and provide more information on how this is actually implemented. Nevertheless, I hope the above example gives you a high level idea on how the copy mechanism is computing the word probabilities during generation.

**Comment by loveritsu929** [comment]
> Hi ! I think what happens (on a high level) is that a new vocab is created in the process. Here is a simple example:
> 
> input: new york is a city in usa .
> question: where is new york located ?
> 
> The original vocab for this question: ["where", "is", "new", "york", "located"]
> 
> let's say we are generating the third word. Ideally, we want to generate the word "new".
> 
> Without copy mechanism, the following could be the softmax output (bad example but illustrative):
> ["where": 0.2, 
> "is":0.2, 
> "new":0.2,
> "york":0.2, 
> "located":0.2]
> 
> With copy mechanism, we do three things here:
> 1) get the attention weights for the words in the input:
> ["new": 0.3,  
> "york": 0.1, 
> "is": 0.1, 
> "a": 0.1, 
> "city": 0.1, 
> "in": 0.1, 
> "usa": 0.2]
> 2) compute the copy probability, say it is 0.8, which means we prefer to copy a word from the input to the generated question
> 3) add the two lists together, weighted by the copy probability. So for example, the word "where" gets a probability of 0.2 * (1-0.8) + 0 * 0.8 = 0.04 (weighted by the not-copy probability which is 1-0.8=0.2, and add 0 * 0.2 because it does not appear in the input). Similarly, the word "new" gets a probability of 0.2 * (1-0.8) + 0.3 * 0.8 = 0.28. The other word probabilities can be computed similarly. 
> 
> Carrying the above process for all words in the input and question, you'll see that the word "new" has the highest probability among the augmented vocabulary (vocab of question + vocab of input) and thus we select this word. 
> 
> It has been some time since I update the repo but I can take a look at the implementation and provide more information on how this is actually implemented. Nevertheless, I hope the above example gives you a high level idea on how the copy mechanism is computing the word probabilities during generation.

Thanks for the enlightening example！

Yes，I agree that on the high level these two probabilities should be “merged” rather than “stitched together” with potential vocab overlap.  I guess I just need to spend more time  figuring out how exactly it is implemented.

I really appreciate your detailed replies. Thank you again for that. :)

**Comment by zichaow** [comment]
Glad you find the above example useful ! let me know if I can help clarify other things.
